tidb: global memory arbitrator mechanism

solotzg opened the pull request:
### First-time contributors' checklist

- [x] I've signed the [**Contributor License Agreement**](https://cla.pingcap.net/pingcap/docs), which is required for the repository owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add documentation for the TiDB global memory arbitrator mode, including:

- Add a Memory arbitrator mode section to configure-memory-usage.md, describing the working principles, configuration methods, monitoring metrics, and best practices for the `standard` and `priority` modes.
- Add descriptions for the following system variables to system-variables.md:
    - `tidb_mem_arbitrator_mode`: sets the global memory management mode for a TiDB instance.
    - `tidb_mem_arbitrator_query_reserved`: sets the amount of memory reserved in advance for a SQL statement before execution.
    - `tidb_mem_arbitrator_soft_limit`: sets the upper limit of memory allocation for a TiDB instance.
    - `tidb_mem_arbitrator_wait_averse`: controls the behavior of SQL statements when waiting for memory resources, including the meanings of the values `0` (default, disabled), `1`, and `nolimit`.
   
### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v9.0 (TiDB 9.0 versions)
- [ ] v8.5 (TiDB 8.5 versions)
- [ ] v8.1 (TiDB 8.1 versions)
- [ ] v7.5 (TiDB 7.5 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)

### What is the related PR or file link(s)?

- Other reference link(s): ref https://github.com/pingcap/tidb/pull/63073 / design doc: [TiDB Global Memory Arbitrator](https://github.com/pingcap/tidb/blob/c2e1dfa992795e778260aa03da52f632742feacf/docs/design/2025-04-15-global-memory-arbitrator.md)

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch
- [ ] Might cause conflicts after applied to another branch

solotzg committed:
Document global memory management framework in TiDB

Added details about the global memory management framework in TiDB, including modes, error handling, monitoring, and practical deployment recommendations.

solotzg committed:
Document new TiDB memory management system variables

Added new system variables related to memory management in TiDB, including details on their scopes, types, default values, and warnings regarding experimental features.

solotzg committed:
Revise TiDB global memory management documentation

Updated the documentation for the global memory management framework in TiDB, including changes to error messages, memory management modes, and monitoring metrics.

solotzg committed:
Update system-variables.md

solotzg committed:
Update memory management mode descriptions in docs

solotzg committed:
Fix formatting and links in memory usage documentation

solotzg committed:
Fix formatting and clarify SQL memory arbitration

solotzg committed:
Fix formatting and indentation in system-variables.md

hfxsd committed:
Apply suggestions from code review

solotzg committed:
Update memory management configuration details

Added configuration file parameters for memory management.

solotzg committed:
Clarify behavior of memory management variable

Correct the phrasing and clarify the behavior of the variable in the context of memory management.

solotzg committed:
Enhance documentation on memory arbitrator modes

joey-yez commented on system-variables.md:
```diff
@@ -3328,6 +3328,71 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 可选值：`disable`，`standard`, `priority`
+- 该变量用于设置 TiDB 实例的全局内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#全局内存管理架构)：
+    - `disable` 为默认模式，[控制行为同上](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入)：内存资源先使用后上报
+    - `standard` 和 `priority` 为先订阅后分配模式：
```
```suggestion
```

solotzg committed:
Update system-variables.md

Co-authored-by: joey-yez <104608045+joey-yez@users.noreply.github.com>

solotzg committed:
Update system-variables.md

Co-authored-by: joey-yez <104608045+joey-yez@users.noreply.github.com>

solotzg committed:
Update system-variables.md

Co-authored-by: joey-yez <104608045+joey-yez@users.noreply.github.com>

solotzg committed:
Update system-variables.md

Co-authored-by: joey-yez <104608045+joey-yez@users.noreply.github.com>

solotzg committed:
Update memory arbitration modes in documentation

Clarified descriptions for memory arbitration modes in TiDB documentation.

solotzg committed:
exchange parts

solotzg committed:
Enhance documentation on TiDB memory management

Refine the explanation of the global memory management framework in TiDB, including details on modes, error handling, and monitoring metrics.

solotzg committed:
Fix formatting and clarify memory arbitration modes

Corrected formatting and clarified explanations for memory arbitration modes.

solotzg committed:
Update memory management descriptions in documentation

solotzg committed:
Update memory arbitration mode documentation

solotzg committed:
Fix links and descriptions for memory management variables

Updated documentation for memory arbitration modes in TiDB.

solotzg committed:
Updated the memory arbitration mode section

solotzg committed:
Update memory arbitration mode documentation

solotzg committed:
Fix typo in memory arbitration section

solotzg commented:
@hfxsd PM has approved this PR. Pls help to merge it.

solotzg committed:
Update memory usage documentation for clarity

Clarify the definition of 'out-of-control' memory in the documentation.

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- 设置为 `standard` 或 `priority` 表示启用内存仲裁模式：
+    - 内存资源的使用按照先订阅后分配的机制，由各个 TiDB 实例中唯一的仲裁者统筹
+    - 预期 TiDB 实例的整体内存使用量不会超过 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的限制，且 [内存占用过高时的报警](#tidb-server-内存占用过高时的报警) 不再生效
+    - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-从-v610-版本开始引入)，[`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 以及 [`max_execution_time`](/system-variables.md#max_execution_time) 的控制行为依旧生效
```
```suggestion
- `disable`：表示禁用内存仲裁模式

- `standard` 或 `priority`：表示启用内存仲裁模式。启用后：
    - 内存资源的使用按照先订阅后分配的机制，由各个 TiDB 实例中唯一的仲裁者统筹
    - 预期 TiDB 实例的整体内存使用量不会超过 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的限制，且 [内存占用过高时的报警](#tidb-server-内存占用过高时的报警) 不再生效
    - 以下系统变量的控制行为依旧有效：
        - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-从-v610-版本开始引入)
        - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)
        - [`max_execution_time`](/system-variables.md#max_execution_time)
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - 预期 TiDB 实例的整体内存使用量不会超过 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的限制，且 [内存占用过高时的报警](#tidb-server-内存占用过高时的报警) 不再生效
+    - [`tidb_mem_oom_action`](/system-variables.md#tidb_mem_oom_action-从-v610-版本开始引入)，[`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 以及 [`max_execution_time`](/system-variables.md#max_execution_time) 的控制行为依旧生效
+
+仲裁者可通过终止 SQL 来回收内存资源（不可终止 `DDL`|`DCL`|`TCL` 类型 SQL），并向客户端返回编号为 `8180` 的错误。错误格式为：`Query execution was stopped by the global memory arbitrator [reason=?, path=?] [conn=?]`：
```
```suggestion
仲裁者可通过终止 SQL 来回收内存资源（不可终止 `DDL`|`DCL`|`TCL` 类型 SQL），并向客户端返回编号为 `8180` 的错误。错误格式为：`Query execution was stopped by the global memory arbitrator [reason=?, path=?] [conn=?]`。相关字段解释如下：
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### `standard` 模式
+
+SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者按先来先服务原则处理订阅请求
```
```suggestion
在 `standard` 模式下，SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者按先来先服务原则处理订阅请求。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### 内存风险控制
+
+当 TiDB 实例的内存使用量达到 `tidb_server_memory_limit` 的 `95%` 这一阈值时，仲裁者开始处理内存风险。如果面临 `OOM` 风险（实际内存使用量短期无法降低到安全线或者实际内存释放速率过小），仲裁者会按顺序（优先级从低到高，内存份额占用量从大到小）强制终止 SQL，返回错误中 `reason` 字段为 `KILL(out-of-memory)`。
```
```suggestion
当 TiDB 实例的内存使用量达到 `tidb_server_memory_limit` 的 `95%` 这一阈值时，仲裁者开始处理内存风险。如果实际内存使用量短期无法降低到安全线或者实际内存释放速率过小，将面临 `OOM` 风险，仲裁者会按顺序（优先级从低到高，内存份额占用量从大到小）强制终止 SQL，返回错误中 `reason` 字段为 `KILL(out-of-memory)`。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### 手动保障内存安全
+
+通过系统变量 [`tidb_mem_arbitrator_soft_limit`](/system-variables.md#tidb_mem_arbitrator_soft_limit-从-v900-版本开始引入) 或 TiDB 配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 可以设置 TiDB 实例中仲裁者的内存份额上限。上限越小，全局内存越安全，但内存资源利用率越低。该变量可用于手动快速收敛内存风险。
```
```suggestion
你可以通过系统变量 [`tidb_mem_arbitrator_soft_limit`](/system-variables.md#tidb_mem_arbitrator_soft_limit-从-v900-版本开始引入) 或 TiDB 配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 设置 TiDB 实例中仲裁者的内存份额上限。上限越小，全局内存越安全，但内存资源利用率越低。该变量可用于手动快速收敛内存风险。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `Mem Quota Arbitration`：内存资源订阅请求处理耗时
+- `Mem Pool Stats`：各类内存池的数量
+- `Runtime Mem Pressure`：内存压力值（实际内存使用量和已分配内存份额的比率）
+- `Waiting Tasks`：排队等待内存分配的各类任务数量
```
```suggestion
- `Work Mode`：TiDB 实例的内存管理模式
- `Arbitration Exec`：仲裁处理各类请求的统计
- `Events`：仲裁模式内各类事件的统计。需要重点关注 `mem-risk`（内存风险）、`oom-risk`（`OOM` 风险）
- `Mem Quota Stats`：各类内存份额占用。需要重点关注 `out-of-control`、`wait-alloc`（订阅等待中的总内存份额）、`mem-inuse`（实际内存使用量）
- `Mem Quota Arbitration`：内存资源订阅请求处理耗时
- `Mem Pool Stats`：各类内存池的数量
- `Runtime Mem Pressure`：内存压力值，即实际内存使用量和已分配内存份额的比率
- `Waiting Tasks`：排队等待内存分配的各类任务数量
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- `MEM_ARBITRATION`：目前为止 SQL 等待内存资源的总耗时
+- `MEM_WAIT_ARBITRATE_START`：当前内存资源订阅请求的开始时间（没有则为 NULL）
+- `MEM_WAIT_ARBITRATE_BYTES`：当前内存资源订阅请求的申请字节数（没有则为 NULL）
```
```suggestion
- `MEM_WAIT_ARBITRATE_START`：当前内存资源订阅请求的开始时间，没有数据则为 NULL
- `MEM_WAIT_ARBITRATE_BYTES`：当前内存资源订阅请求的申请字节数，没有数据则为 NULL
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `AVG_MEM_ARBITRATION`：平均 SQL 等待内存资源耗时
+- `MAX_MEM_ARBITRATION`：最大 SQL 等待内存资源耗时
+
+### 实践
```
```suggestion
### 最佳实践
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+系统变量 `tidb_server_memory_limit` 的默认值 (`80%`) 较小，启用内存仲裁模式后建议设置为 `95%`。
+
+部署单节点 TiDB 场景
```
```suggestion
在部署单节点 TiDB 的场景中：
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 按需设置 TiDB 实例内存份额上限
+    - 按需分发 SQL 到不同组，并按照上述不同模式处理
+
+保障重要 SQL 执行
```
```suggestion
可以采取以下措施保障重要 SQL 执行：
```

hfxsd commented on system-variables.md:
```diff
@@ -3328,6 +3328,69 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 该变量用于设置 TiDB 实例的全局内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)：
+    - `disable` 表示禁用内存仲裁模式，保持内存资源 [先使用后上报](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的机制。
+    - `standard` 表示启用标准内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。当订阅内存资源失败后，SQL 终止执行。
+    - `priority` 表示启用基于优先级的内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。TiDB 根据 SQL 的 [资源组优先级](/information-schema/information-schema-resource-groups.md) 处理内存资源订阅请求。
```
```suggestion
- 该变量用于设置 TiDB 实例的全局内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。有以下取值：
    - `disable`：表示禁用内存仲裁模式，保持内存资源[先使用后上报](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入)的机制。
    - `standard`：表示启用标准内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。当订阅内存资源失败后，SQL 终止执行。
    - `priority`：表示启用基于优先级的内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。TiDB 根据 SQL 的 [资源组优先级](/information-schema/information-schema-resource-groups.md) 处理内存资源订阅请求。
```

hfxsd commented on system-variables.md:
```diff
@@ -3328,6 +3328,69 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 类型：字符串
+- 默认值：`0`
+- 可选值：`0`，浮点数 `(0, 1]`，整数 `(1, 9223372036854775807]`
+- 当启用内存仲裁模式后，该变量设置 TiDB 实例的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```
```suggestion
- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量设置 TiDB 实例的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```

hfxsd committed:
Apply suggestions from code review

solotzg commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,144 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- TiDB 实例内存超限后 SQL 随机被终止
+- 内存资源先使用后上报，且 SQL 间存在信息孤岛，难以全局控制
+- 内存压力较大时 Golang 垃圾回收 (GC) 开销过高，严重时则引起内存溢出 (OOM) 问题
```
> “随机”还是“随即”？
> 

随机

solotzg committed:
Apply suggestion from @hfxsd

Co-authored-by: xixirangrang <hfxsd@hotmail.com>

solotzg committed:
Apply suggestion from @hfxsd

Co-authored-by: xixirangrang <hfxsd@hotmail.com>

hfxsd committed:
Update configure-memory-usage.md

hfxsd committed:
Apply suggestion from @hfxsd

hfxsd committed:
Apply suggestion from @hfxsd

hfxsd committed:
Merge remote-tracking branch 'upstream/master' into solotzg-patch-1

hfxsd committed:
Apply suggestions from code review

Copilot committed:
system-variables: add explanation for value 0 of tidb_mem_arbitrator_wait_averse

Agent-Logs-Url: https://github.com/pingcap/docs-cn/sessions/a019b30c-a16a-48b0-98e0-61848219236f

Co-authored-by: solotzg <9016847+solotzg@users.noreply.github.com>

Copilot committed:
Merge branch 'master' of https://github.com/pingcap/docs-cn into solotzg-patch-1
Co-authored-by: solotzg <9016847+solotzg@users.noreply.github.com>

solotzg committed:
Revise warning for experimental memory arbitration feature

Updated warning message regarding experimental feature for memory arbitration mode.

solotzg committed:
Apply suggestion from @solotzg

hfxsd committed:
Merge remote-tracking branch 'upstream/master' into solotzg-patch-1

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- TiDB 实例内存超限后，SQL 随机被终止
+- 内存资源先使用后上报，且 SQL 间存在信息孤岛，难以全局控制
+- 内存压力较大时，Golang 垃圾回收 (Garbage Collection, GC) 开销过高，严重时导致内存溢出 (OOM) 问题
```
```suggestion
- TiDB 实例内存使用量超过限制时，TiDB 可能随机终止正在运行的 SQL 语句。
- 内存资源采用先使用后上报的机制，且不同 SQL 之间的内存使用信息相互隔离，TiDB 难以从实例级别统一调度和控制内存资源。
- 内存压力较大时，Go 垃圾回收 (Garbage Collection, GC) 的开销显著升高，严重时可能导致内存溢出 (OOM) 问题。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- 内存资源先使用后上报，且 SQL 间存在信息孤岛，难以全局控制
+- 内存压力较大时，Golang 垃圾回收 (Garbage Collection, GC) 开销过高，严重时导致内存溢出 (OOM) 问题
+
+从 v9.0.0 开始，TiDB 引入了内存仲裁模式这种新机制，通过自上向下地集中化管理和调度内存资源来解决上述问题。
```
```suggestion
从 v9.0.0 开始，TiDB 引入了内存仲裁模式。该模式在每个 TiDB 实例内引入全局内存仲裁者，通过自上向下地集中管理和调度该实例的内存资源，缓解上述问题。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+        - [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)
+        - [`max_execution_time`](/system-variables.md#max_execution_time)
+
+仲裁者可通过终止 SQL 来回收内存资源（不可终止 `DDL`|`DCL`|`TCL` 类型 SQL），并向客户端返回编号为 `8180` 的错误。错误格式为：`Query execution was stopped by the global memory arbitrator [reason=?, path=?] [conn=?]`。相关字段解释如下：
```
```suggestion
当内存资源不足或存在 OOM 风险时，仲裁者可以通过终止 SQL 来回收内存资源，但不会终止 `DDL`、`DCL` 和 `TCL` 类型的 SQL。被终止的 SQL 会向客户端返回编号为 `8180` 的错误。错误格式为：`Query execution was stopped by the global memory arbitrator [reason=?, path=?] [conn=?]`。相关字段解释如下：
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- `conn`：连接（会话）ID
+- `reason`：终止 SQL 的具体原因
+- `path`：SQL 被终止的阶段，无 `path` 字段则默认为执行阶段
```
```suggestion
- `path`：SQL 被终止的阶段（如果错误中没有 `path` 字段，表示 SQL 在执行阶段被终止）
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - `ParseSQL`：解析
+    - `CompilePlan`：编译执行计划
+
+内存仲裁模式将 TiDB 实例的内存资源池化，其中 `tidb_server_memory_limit` 总共划分为 4 个部分，对应 `Mem Quota Stats` 中监控指标：
```
```suggestion
内存仲裁模式将每个 TiDB 实例内的可仲裁内存资源池化，其中`tidb_server_memory_limit` 对应的内存资源分为以下 4 类，可通过 `Mem Quota Stats` 监控指标查看：
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `out-of-control`：不受仲裁者控制的内存份额，包括但不限于：
+    - 不安全或禁用的内存
+    - 等待垃圾回收的内存
+    - Golang runtime 的内存占用
```
```suggestion
    - Go runtime 的内存占用
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - Golang runtime 的内存占用
+    - 其他未被追踪到的内存使用
+
+`out-of-control` 属于关键风险指标，由仲裁者动态计算得出，最小值为 `tidb_server_memory_limit - tidb_mem_arbitrator_soft_limit`。仲裁者会尽可能准确地量化这种有风险的内存份额，在分配内存时避免占用，从而保障全局内存安全。
```
```suggestion
`out-of-control` 是需要重点关注的风险指标，由仲裁者动态计算得出，最小值为 `tidb_server_memory_limit - tidb_mem_arbitrator_soft_limit`。仲裁者会尽可能准确地量化这类有风险的内存份额，并在分配内存时为其预留空间，从而降低 TiDB 实例发生 OOM 的风险。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### `standard` 模式
+
+在 `standard` 模式下，SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者按先来先服务原则处理订阅请求。
```
```suggestion
在 `standard` 模式下，SQL 运行过程中会动态向仲裁者订阅内存资源，并在资源不足时阻塞等待。仲裁者按先来先服务 (FIFO) 的原则处理订阅请求。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+在 `standard` 模式下，SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者按先来先服务原则处理订阅请求。
+
+- 解析 SQL 或编译执行计划时，需要订阅的内存份额由 TiDB 估算，与 SQL 关键字数量成正比
+- 如果全局内存资源不足，仲裁者令请求失败并终止 SQL，返回错误中 `reason` 字段为 `CANCEL(out-of-quota & standard-mode)`
```
```suggestion
- 解析 SQL 或编译执行计划时，需要订阅的内存份额由 TiDB 估算，与 SQL 关键字数量成正比。
- 如果全局内存资源不足，仲裁者会令订阅请求失败并终止 SQL，返回错误中的 `reason` 字段为 `CANCEL(out-of-quota & standard-mode)`。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### `priority` 模式
+
+在 `priority` 模式下，SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者根据 SQL 的[资源组优先级](/information-schema/information-schema-resource-groups.md) (`LOW | MEDIUM | HIGH`) 处理订阅请求。
```
```suggestion
在 `priority` 模式下，SQL 运行过程中会动态向仲裁者订阅内存资源，并在资源不足时阻塞等待。仲裁者根据 SQL 所属[资源组优先级](/information-schema/information-schema-resource-groups.md) (`LOW | MEDIUM | HIGH`) 处理订阅请求。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+在 `priority` 模式下，SQL 运行过程中会动态地向仲裁者订阅内存资源并阻塞式等待，仲裁者根据 SQL 的[资源组优先级](/information-schema/information-schema-resource-groups.md) (`LOW | MEDIUM | HIGH`) 处理订阅请求。
+
+- 解析 SQL 或编译执行计划时，需要订阅的内存份额由 TiDB 估算，与 SQL 关键字数量成正比。不同于 `standard` 模式，在 `priority` 模式下，除非面临 `OOM` 风险，否则订阅失败后不会立刻终止 SQL。
+- 所有订阅请求按照优先级从高到低排队等待资源，同级别请求按照发起顺序排序。
```
```suggestion
- 仲裁者按照优先级从高到低处理订阅请求。同一优先级的请求按发起顺序排队。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- 所有订阅请求按照优先级从高到低排队等待资源，同级别请求按照发起顺序排序。
+- 全局内存资源不足时：
+    - 仲裁者按顺序（优先级从低到高，内存份额占用量从大到小）终止低优先级 SQL，回收资源来满足高优先级 SQL。返回错误中 `reason` 字段为 `CANCEL(out-of-quota & priority-mode)`。
+    - 如果没有可终止的 SQL，则等待现有 SQL 执行完后释放资源。
```
```suggestion
    - 如果没有可终止的 SQL，订阅请求会继续等待，直到已有 SQL 执行完成并释放资源。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+### 内存风险控制
+
+当 TiDB 实例的内存使用量达到 `tidb_server_memory_limit` 的 `95%` 这一阈值时，仲裁者开始处理内存风险。如果实际内存使用量短期无法降低到安全线或者实际内存释放速率过小，将面临 `OOM` 风险，仲裁者会按顺序（优先级从低到高，内存份额占用量从大到小）强制终止 SQL，返回错误中 `reason` 字段为 `KILL(out-of-memory)`。
```
```suggestion
当 TiDB 实例的内存使用量达到 `tidb_server_memory_limit` 的 `95%` 这一阈值时，仲裁者开始处理内存风险。如果实际内存使用量短期无法降低到安全线或者实际内存释放速率过小，TiDB 实例将面临 `OOM` 风险，仲裁者会按顺序（优先级从低到高，内存份额占用量从大到小）强制终止 SQL，返回错误中 `reason` 字段为 `KILL(out-of-memory)`。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- `Work Mode`：TiDB 实例的内存管理模式
+- `Arbitration Exec`：仲裁处理各类请求的统计
+- `Events`：仲裁模式内各类事件的统计。需要重点关注 `mem-risk`（内存风险）、`oom-risk`（`OOM` 风险）
```
```suggestion
- `Events`：仲裁模式下各类事件的统计。需要重点关注 `mem-risk`（内存风险）和`oom-risk`（`OOM` 风险）
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `Work Mode`：TiDB 实例的内存管理模式
+- `Arbitration Exec`：仲裁处理各类请求的统计
+- `Events`：仲裁模式内各类事件的统计。需要重点关注 `mem-risk`（内存风险）、`oom-risk`（`OOM` 风险）
+- `Mem Quota Stats`：各类内存份额占用。需要重点关注 `out-of-control`、`wait-alloc`（订阅等待中的总内存份额）、`mem-inuse`（实际内存使用量）
```
```suggestion
- `Mem Quota Stats`：各类内存份额占用。需要重点关注 `out-of-control`、`wait-alloc`（等待分配的总内存份额）、`mem-inuse`（实际内存使用量）
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `Runtime Mem Pressure`：内存压力值，即实际内存使用量和已分配内存份额的比率
+- `Waiting Tasks`：排队等待内存分配的各类任务数量
+
+[`SLOW_QUERY`](/information-schema/information-schema-slow-query.md) 新增 `Mem_arbitration` 字段表示 SQL 等待内存资源的总耗时。[TiDB Dashboard 慢查询页面](/dashboard/dashboard-slow-query.md) 中的 `Mem Arbitration` 列也会显示该信息。
```
```suggestion
[`SLOW_QUERY`](/information-schema/information-schema-slow-query.md) 新增 `Mem_arbitration` 字段，表示 SQL 等待内存资源的总耗时。[TiDB Dashboard 慢查询页面](/dashboard/dashboard-slow-query.md) 中的 `Mem Arbitration` 列也会显示该信息。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- `MEM_ARBITRATION`：目前为止 SQL 等待内存资源的总耗时
+- `MEM_WAIT_ARBITRATE_START`：当前内存资源订阅请求的开始时间，没有数据则为 NULL
+- `MEM_WAIT_ARBITRATE_BYTES`：当前内存资源订阅请求的申请字节数，没有数据则为 NULL
```
```suggestion
- `MEM_WAIT_ARBITRATE_START`：当前内存资源订阅请求的开始时间，如果当前没有等待中的订阅请求则为 NULL
- `MEM_WAIT_ARBITRATE_BYTES`：当前内存资源订阅请求的申请字节数，如果当前没有等待中的订阅请求则为 NULL
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `MEM_WAIT_ARBITRATE_START`：当前内存资源订阅请求的开始时间，没有数据则为 NULL
+- `MEM_WAIT_ARBITRATE_BYTES`：当前内存资源订阅请求的申请字节数，没有数据则为 NULL
+
+[`Expensive query`](/identify-expensive-queries.md) 新增字段 `mem_arbitration`，用于记录 SQL 等待资源耗时和当前订阅请求的信息，例如：
```
```suggestion
[`Expensive query`](/identify-expensive-queries.md) 新增字段 `mem_arbitration`，用于记录 SQL 等待内存资源的耗时和当前订阅请求的信息，例如：
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+[`STATEMENTS_SUMMARY`](/statement-summary-tables.md)：表 `statements_summary`、`statements_summary_history`、`cluster_statements_summary`、`cluster_statements_summary_history` 新增以下字段，[TiDB Dashboard SQL 语句分析执行详情页面](/dashboard/dashboard-statement-list.md) 中的 `Mean Mem Arbitration` 列也会显示该信息：
+
+- `AVG_MEM_ARBITRATION`：平均 SQL 等待内存资源耗时
+- `MAX_MEM_ARBITRATION`：最大 SQL 等待内存资源耗时
```
```suggestion
- `AVG_MEM_ARBITRATION`：SQL 等待内存资源的平均耗时
- `MAX_MEM_ARBITRATION`：SQL 等待内存资源的最大耗时
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+在部署单节点 TiDB 的场景中：
+
+- 开启 `priority` 模式：绑定 OLTP 相关或重要 SQL 到高优先级，按需绑定其他 SQL 到中或低优先级
+- 开启 `standard` 模式：遇到 `8180` 错误，需等待并重试 SQL
```
```suggestion
- 如果开启 `priority` 模式，建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组
- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在等待一段时间后重试该 SQL
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 类型：枚举型
+- 默认值：`0`
+- 可选值：`0`，`1`，`nolimit`
+- 当启用内存仲裁模式后，该变量控制 SQL 等待内存资源的相关行为，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。
```
```suggestion
- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量用于控制 SQL 等待内存资源时的行为，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 当启用内存仲裁模式后，该变量控制 SQL 等待内存资源的相关行为，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。
+    - `0`（默认值）：表示禁用该功能，变量不生效
+    - `1`：`priority` 模式下，SQL 订阅内存资源时自动绑定高优先级，全局内存资源不足时就终止执行而非阻塞式等待
+    - `nolimit`：SQL 使用内存资源不受限制，可能导致 TiDB 实例内存风险
```
```suggestion
    - `1`：仅在 `priority` 模式下生效。SQL 订阅内存资源时自动绑定为高优先级；当全局内存资源不足时，SQL 会终止执行，而不是阻塞等待。
    - `nolimit`：SQL 的内存使用不受仲裁者限制。该取值可能增加 TiDB 实例发生 OOM 的风险。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 类型：字符串
+- 默认值：`0`
+- 可选值：`0`，浮点数 `(0, 1]`，整数 `(1, 9223372036854775807]`
+- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量设置 TiDB 实例的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```
```suggestion
- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量控制 TiDB 实例中仲裁者可分配的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 可选值：`0`，浮点数 `(0, 1]`，整数 `(1, 9223372036854775807]`
+- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量设置 TiDB 实例的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
+    - `0`：默认资源份额上限为 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 值的 `95%`
+    - 浮点数 `(0, 1]`：指定比率，资源份额上限为 `tidb_mem_arbitrator_soft_limit *` [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入)
```
```suggestion
    - 浮点数 `(0, 1]`：指定资源份额上限相对于 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的比例。例如，`0.8` 表示资源份额上限为 `tidb_server_memory_limit * 0.8`。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 默认值：`0`
+- 单位：字节
+- 范围：`[0, 9223372036854775807]`
+- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量设置 SQL 执行前预先订阅指定数量的内存资源份额，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```
```suggestion
- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量控制 SQL 执行前向内存仲裁者预先订阅的内存资源份额，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：枚举型
+- 默认值：`disable`
+- 可选值：`disable`，`standard`, `priority`
```
```suggestion
- 可选值：`disable`、`standard`、`priority`
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 该变量用于设置 TiDB 实例的全局内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。有以下取值：
+    - `disable`（默认值）：表示禁用内存仲裁模式，保持内存资源[先使用后上报](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入)的机制。
+    - `standard`：表示启用标准内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。当订阅内存资源失败后，SQL 终止执行。
+    - `priority`：表示启用基于优先级的内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，成功后再分配内存资源。TiDB 根据 SQL 的[资源组优先级](/information-schema/information-schema-resource-groups.md)处理内存资源订阅请求。
```
```suggestion
    - `standard`：表示启用标准内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，订阅成功后再分配内存资源。如果订阅失败，SQL 终止执行。
    - `priority`：表示启用基于优先级的内存仲裁模式。SQL 需要使用内存资源时，先向内存仲裁者进行订阅，订阅成功后再分配内存资源。TiDB 根据 SQL 的[资源组优先级](/information-schema/information-schema-resource-groups.md)处理内存资源订阅请求。
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 类型：枚举型
+- 默认值：`disable`
+- 可选值：`disable`，`standard`, `priority`
+- 该变量用于设置 TiDB 实例的全局内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。有以下取值：
```
```suggestion
- 该变量用于设置 TiDB 实例的内存管理模式，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。有以下取值：
```

qiancai commented on system-variables.md:
```diff
@@ -3417,6 +3417,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 是否受 Hint [SET_VAR](/optimizer-hints.md#set_varvar_namevar_value) 控制：否
+- 类型：字符串
+- 默认值：`0`
+- 可选值：`0`，浮点数 `(0, 1]`，整数 `(1, 9223372036854775807]`
```
这个变量取值也支持 `auto`？

solotzg replied:
No

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+在部署多节点 TiDB 的场景中：
+
+- 如果开启 `standard` 模式，遇到 `8180` 错误，则重试 SQL 到其他 TiDB 节点
```
```suggestion
- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在其他 TiDB 节点重试该 SQL

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - 为 OLTP 相关或重要 SQL 绑定高优先级，按需绑定其他 SQL 到中或低优先级
+    - 通过 `max_execution_time` 限制 SQL 最大执行时间
+    - 遇到超时或 `8180` 错误，则重试 SQL 到其他 TiDB 节点
+    - 可通过设置 `tidb_mem_arbitrator_wait_averse` 使 SQL 尽快重试到内存资源充足的节点
```
```suggestion
    - 建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组
    - 通过 `max_execution_time` 限制 SQL 最大执行时间
    - 遇到超时或 `8180` 错误，建议在其他 TiDB 节点重试该 SQL
    - 如果希望 SQL 在内存资源不足时尽快失败并重试到其他节点，可以设置 `tidb_mem_arbitrator_wait_averse`
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- TiDB 实例分组
+    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_mode` 设置 TiDB 实例内存管理模式
+    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 按需设置 TiDB 实例内存份额上限
+    - 按需分发 SQL 到不同组，并按照上述不同模式处理
```
```suggestion
    - 按业务需求将 SQL 分发到不同实例组，并根据各实例组的仲裁模式处理 SQL 失败或重试
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- 通过 `tidb_mem_arbitrator_query_reserved` 保障 SQL 的内存资源隔离性
+- 绑定高优先级的资源组
+- 设置 `tidb_mem_quota_query` 为较大值以免 SQL 执行被中断
+- 保底方案为设置 `tidb_mem_arbitrator_wait_averse` 为 `nolimit` 并承担相关风险
```
```suggestion
- 通过 `tidb_mem_arbitrator_query_reserved` 为重要 SQL 预先订阅内存份额，增强其内存资源隔离性。
- 将重要 SQL 绑定到高优先级资源组。
- 设置 `tidb_mem_quota_query` 为较大的值，以降低单条 SQL 因超过内存配额而被中断的概率。
- 在能够接受 OOM 风险的情况下，将 `tidb_mem_arbitrator_wait_averse` 设置为 `nolimit`，使相关 SQL 跳过内存仲裁限制。
```

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+>
+> 该功能为实验特性，不建议在生产环境中使用。该功能可能会在未事先通知的情况下发生变化或删除。如果发现 bug，请在 GitHub 上提 [issue](https://github.com/pingcap/tidb/issues) 反馈。
+
+你可以通过系统变量 [`tidb_mem_arbitrator_mode`](/system-variables.md#tidb_mem_arbitrator_mode-从-v900-版本开始引入) 或 TiDB 配置文件参数 `instance.tidb_mem_arbitrator_mode` 开启内存仲裁模式。
```
https://docs.pingcap.com/zh/tidb/dev/tidb-configuration-file/ 文档需要加上这个配置项吗

solotzg replied:
暂时不需要，等 feature 稳定后再统一改到其他文档中

qiancai commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- `disable`：表示禁用内存仲裁模式
+
+- `standard` 或 `priority`：表示启用内存仲裁模式。启用后：
+    - 内存资源的使用按照先订阅后分配的机制，由各个 TiDB 实例中唯一的仲裁者统筹
```
```suggestion
    - 内存资源的使用按照先订阅后分配的机制，由每个 TiDB 实例内的内存仲裁者统一调度
```

solotzg commented:
@copilot apply

Copilot committed:
Apply @qiancai review suggestions to configure-memory-usage.md and system-variables.md

Agent-Logs-Url: https://github.com/pingcap/docs-cn/sessions/623b0870-0611-40e2-8aaa-b4eead1118b1

Co-authored-by: solotzg <9016847+solotzg@users.noreply.github.com>

hfxsd committed:
Merge remote-tracking branch 'upstream/master' into solotzg-patch-1

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+- `standard` 或 `priority`：表示启用内存仲裁模式。启用后：
+    - 内存资源的使用按照先订阅后分配的机制，由每个 TiDB 实例内的内存仲裁者统一调度
+    - 预期 TiDB 实例的整体内存使用量不会超过 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的限制，且[内存占用过高时的报警](#tidb-server-内存占用过高时的报警)不再生效
```
```suggestion
    - 内存资源的使用按照先订阅后分配的机制，由每个 TiDB 实例内的内存仲裁者统一调度。
    - 预期 TiDB 实例的整体内存使用量不会超过 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的限制，且[内存占用过高时的报警](#tidb-server-内存占用过高时的报警)不再生效。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+
+在部署多节点 TiDB 的场景中：
+
+- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在其他 TiDB 节点重试该 SQL
```
```suggestion
- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在其他 TiDB 节点重试该 SQL。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+在部署单节点 TiDB 的场景中：
+
+- 如果开启 `priority` 模式，建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组
+- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在等待一段时间后重试该 SQL
```
```suggestion
- 如果开启 `priority` 模式，建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组。
- 如果开启 `standard` 模式，当 SQL 返回 `8180` 错误时，建议在等待一段时间后重试该 SQL。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+    - 建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组
+    - 通过 `max_execution_time` 限制 SQL 最大执行时间
+    - 遇到超时或 `8180` 错误，建议在其他 TiDB 节点重试该 SQL
+    - 如果希望 SQL 在内存资源不足时尽快失败并重试到其他节点，可以设置 `tidb_mem_arbitrator_wait_averse`
```
```suggestion
    - 建议将 OLTP 相关 SQL 或业务关键 SQL 绑定到高优先级资源组，并按需将其他 SQL 绑定到中优先级或低优先级资源组。
    - 通过 `max_execution_time` 限制 SQL 最大执行时间。
    - 遇到超时或 `8180` 错误，建议在其他 TiDB 节点重试该 SQL。
    - 如果希望 SQL 在内存资源不足时尽快失败并重试到其他节点，可以设置 `tidb_mem_arbitrator_wait_averse`。
```

hfxsd commented on configure-memory-usage.md:
```diff
@@ -231,3 +229,145 @@ Golang 自 Go 1.19 版本开始引入 [`GOMEMLIMIT`](https://pkg.go.dev/runtime@
+- TiDB 实例分组
+    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_mode` 设置 TiDB 实例内存管理模式
+    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 按需设置 TiDB 实例内存份额上限
+    - 按业务需求将 SQL 分发到不同实例组，并根据各实例组的仲裁模式处理 SQL 失败或重试
```
```suggestion
    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_mode` 设置 TiDB 实例内存管理模式。
    - 各组内通过配置文件参数 `instance.tidb_mem_arbitrator_soft_limit` 按需设置 TiDB 实例内存份额上限。
    - 按业务需求将 SQL 分发到不同实例组，并根据各实例组的仲裁模式处理 SQL 失败或重试。
```

hfxsd committed:
Apply suggestions from code review

hfxsd commented on system-variables.md:
```diff
@@ -3426,6 +3426,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量控制 TiDB 实例中仲裁者可分配的内存资源份额上限，详见 [TiDB 内存控制](/configure-memory-usage.md#手动保障内存安全)。
+    - `0`：默认资源份额上限为 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 值的 `95%`
+    - 浮点数 `(0, 1]`：指定资源份额上限相对于 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的比例。例如，`0.8` 表示资源份额上限为 `tidb_server_memory_limit * 0.8`。
+    - 整数 `(1, 9223372036854775807]`：指定字节数
```
```suggestion
    - `0`：默认资源份额上限为 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 值的 `95%`。
    - 浮点数 `(0, 1]`：指定资源份额上限相对于 [`tidb_server_memory_limit`](/system-variables.md#tidb_server_memory_limit-从-v640-版本开始引入) 的比例。例如，`0.8` 表示资源份额上限为 `tidb_server_memory_limit * 0.8`。
    - 整数 `(1, 9223372036854775807]`：指定字节数。
```

hfxsd commented on system-variables.md:
```diff
@@ -3426,6 +3426,70 @@ v5.0 后，用户仍可以单独修改以上系统变量（会有废弃警告）
+- 默认值：`0`
+- 可选值：`0`，`1`，`nolimit`
+- 当启用[内存仲裁模式](/configure-memory-usage.md#内存仲裁模式)后，该变量用于控制 SQL 等待内存资源时的行为，详见 [TiDB 内存控制](/configure-memory-usage.md#内存仲裁模式)。
+    - `0`（默认值）：表示禁用该功能，变量不生效
```
```suggestion
    - `0`（默认值）：表示禁用该功能，变量不生效。
```

hfxsd replied:
```suggestion
    - `0`（默认值）：表示禁用该功能，变量不生效。
```

hfxsd committed:
Apply suggestions from code review

hfxsd commented:
/retest

hfxsd committed:
Revert "Apply suggestions from code review"

This reverts commit f0d7c9b16fe0f744fc5812ddb5301c673151377a.

hfxsd committed:
Update system-variables.md

hfxsd committed:
Revert "Update system-variables.md"

This reverts commit 1061c656d756eace8517d3d99829eabe25873a32.

hfxsd commented:
/approve
